src/firefly/js/visualize/ui/CatalogSelectViewPanel.jsx

```diff
@@ -0,0 +1,623 @@
+        }
+    }
+
+    /**
+     * Fetch master catalog table with search id = 'irsaCatalogMasterTable'
+     * 'tableModel.tableData.data' returned in the promise is an array of elements which value is refered to the ith column name
+     * from tableModel.tableData.columns
+     * result is set in the state as 'master' object {catmaster, cols}
+     * @see setMaster
+     * @returns {Void} fetch master table and once fetched, set state with it
+     */
+    loadMasterCatalogTable() {
+
+        const request = {id: 'irsaCatalogMasterTable'}; //Fetch master table
+        TblUtil.doFetchTable(request).then((tableModel) => {
+
+            var data = tableModel.tableData.data;
+
+            cols = tableModel.tableData.columns;
+
+            var projects = [], subprojects = {}, option = [];
+            var i;
+            for (i = 0; i < data.length; i++) {
+                if (projects.indexOf(data[i][0]) < 0) {
+                    projects.push(data[i][0]);
+                }
+            }
+            for (i = 0; i < projects.length; i++) {
+                option = [];
+                const tmp = [];
+                data.forEach((e) => {
+                    if (e[0] === projects[i]) {
+                        if (option.indexOf(e[1]) < 0) {
+                            option.push(e[1]);
+                            var obj = {
+                                label: e[1],
+                                value: e[1],
+                                proj: projects[i]
+                            };
+                            tmp.push(obj);
+                        }
+                    }
+                });
+                subprojects[projects[i]] = tmp;
+
+
+            }
+            var o = 0;
+            for (i = 0; i < projects.length; i++) {
+                var opt = subprojects[projects[i]];
+                const tmp = [];
+                for (o = 0; o < opt.length; o++) {
+                    var myobj = {};
+                    myobj.project = projects[i];
+                    myobj.subproject = opt[o];
+                    var tab = data.filter((e) => {
+                        if (e[0] === projects[i] && e[1] === opt[o].value) {
+                            return e;
+                        }
+                    });
+                    myobj.option = tab.map((e) => {
+                        if (e[0] === projects[i] && e[1] === opt[o].value) {
+                            var rObj = {};
+                            rObj.label = e[2];
+                            rObj.value = e[4];
+                            rObj.proj = e[1];
+                            rObj.cat = e;
+                            return rObj;
+                        }
+                    });
+                    tmp.push(myobj);
+                }
+                catmaster[i] = {
+                    catalogs: tmp,
+                    project: projects[i],
+                    subproject: opt
+                };
+            }
```

Comment:
- You could do a map from `projects` to `catmaster`
- `myobj` could be produced by a `reduce` on `subprojects[projects[i]]`